### PR TITLE
Replaced page column with type

### PR DESCRIPTION
**Pull request by allouis**
refs #10922

Currently we have a `page` column which is a non-nullable boolean column which defaults to `false`

We want to replace this column with a `type` column,
non-nullable string which defaults to `'post'` and has validations to ensure it is obly ever set to `'page'` or `'post'`

- [x] Remove `page` column from schema
- [x] Add `type` column to schema
- [x] Add migration for copying data from `page` -> `type` column
- [x] Add migrations for dropping the `page` column
- [x] Update the model layer to reflect changes
- [x] Update the serialiser layer to reflect changes
- [x] :bug: Fix v2 api responding with 400 when requestign a page from the posts api or a post from the pages api.
- [x] Updated to work with v0.1
- [x] Fixed down migrations

### CODE REVIEW

**Comment by allouis** [comment]
This is missing updates to the v0.1 API. And probably a helluva lot more! :ghost:

File: core/server/models/post.js [hunk 1]
```diff
@@ -621,9 +621,9 @@ Post = ghostBookshelf.Model.extend({
                 options.staticPages = _.includes(['true', '1'], options.staticPages);
             }
 
-            filter = `page:${options.staticPages}`;
```
Comment (gargol) [inline core/server/models/post.js]: Can't wait to drop this shit in v3 :fire:

**Comment by gargol** [comment]
@allouis the direction looks good. like tiny modification approach :+1:  

Was wondering about your thoughts on migrating configurations in frontend for [routing](https://github.com/TryGhost/Ghost/blob/df7e64f/core/frontend/services/routing/config/v0.1.js#L39). There was a change done to URL service in this commit https://github.com/TryGhost/Ghost/pull/10944/commits/d63d9995048e8a7bb6bfd2e6581436e31fcef821 , but it's more of a necessity as it bypasses controller serialization layer and queries the model directly. Routing config state is different as it behaves as frontend should - goes through API layer and parameters will be transformed in the serialization layer.
The question is: should we modify routing config to use type instead even though we don't have to? 
My opinion is to leave it as is so that it continues to behave like v01 API client unless there's a clear benefit of using 'type' there.

**Comment by allouis** [comment]
Yo @gargol :ghost: 

> My opinion is to leave it as is so that it continues to behave like v01 API client unless there's a clear benefit of using 'type' there.

The reason that had to be updated, is because we're removing the `page` column completely so those queries will error unless we update it to use type.

Ideally they would not use the model layer directly, but that change is outside the scope of this PR

File: core/server/data/migrations/versions/3.0/1-add-type-column.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+const common = require('../../../../lib/common');
+const commands = require('../../../schema').commands;
+
+const createLog = type => msg => common.logging[type](msg);
+
+function createColumnMigration({table, column, dbIsInCorrectState, operation}) {
```
Comment (gargol) [inline core/server/data/migrations/versions/3.0/1-add-type-column.js]: I see this method came from `1-add-type-column-to-integrations` :wink: Like it! Think the next time we use 'createColumnMigration' in our migrations we can start thinking of extracting it into some kind of 'migrations-utils' :+1:

File: core/server/data/migrations/versions/3.0/1-add-type-column.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    return function columnMigrations({transacting}) {
+        return transacting.schema.hasColumn(table, column)
+            .then(dbIsInCorrectState)
+            .then((isInCorrectState) => {
+                const log = createLog(isInCorrectState ? 'warn' : 'info');
+
+                log(`Adding ${column} column to ${table} table`);
```
Comment (gargol) [inline core/server/data/migrations/versions/3.0/1-add-type-column.js]: The `Adding` bit won't be always true here as `down` migration is `Dropping`

Comment (gargol) [reply core/server/data/migrations/versions/3.0/1-add-type-column.js]: Also, think it's a little misleading message. When `isCorrectState` is `false` we "skip adding" the column altogether, so would rather give that on the output :thinking:

File: core/server/data/migrations/versions/3.0/3-remove-page-column.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+const createLog = type => msg => common.logging[type](msg);
+
+function createColumnMigration({table, column, dbIsInCorrectState, operation}) {
+    return function columnMigrations(options) {
+        return options.transacting.schema.hasColumn(table, column)
+            .then(dbIsInCorrectState)
+            .then((isInCorrectState) => {
+                const log = createLog(isInCorrectState ? 'warn' : 'info');
+
+                log(`Adding ${column} column to ${table} table`);
```
Comment (gargol) [inline core/server/data/migrations/versions/3.0/3-remove-page-column.js]: Same issue here as in [1-add-type-column.js](https://github.com/TryGhost/Ghost/pull/10944/commits/516e92c89e6472db4af17b7cd09684bac3034564#r307211937)

File: core/server/data/migrations/versions/3.0/2-populate-type-column.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                    common.logging.warn(
+                        `Table '${tableName}' does not have column '${toColumn}'`
+                    );
+                }
+                if (!fromColumnExists || !toColumnExists) {
+                    return;
+                }
+
+                common.logging.info(
+                    `Updating ${tableName}, setting ${fromValue} in ${fromColumn} to ${toValue} in ${toColumn}`
```
Comment (gargol) [inline core/server/data/migrations/versions/3.0/2-populate-type-column.js]: This log message is confusing. Currently, it produces this output:
```
Updating posts, setting true in page to page in type
```
 
Maybe something following would make it clearer:
```
`Updating ${tableName}, setting "${toValue}" value in "${toColumn}" column from "${fromValue}" value in "${fromColumn}" column`
```
Then an example output would read:
```
Updating posts, setting "page" value in "type" column from  "true" value in "page" column
```

**Comment by gargol** [comment]
@allouis when testing 'down' migrations the rollback command exited with error:
```
~> knex-migrator rollback --v 2.25  --force
[2019-07-25 14:09:23] ERROR Cannot read property 'schema' of undefined
```

With what command did you test it?

File: core/server/data/migrations/versions/3.0/1-add-type-column.js (hunk 2 above)
Comment (allouis) [reply core/server/data/migrations/versions/3.0/1-add-type-column.js]: ;) you found it! Yeah that sounds good to me :)

File: core/server/data/migrations/versions/3.0/1-add-type-column.js (hunk 3 above)
Comment (allouis) [reply core/server/data/migrations/versions/3.0/1-add-type-column.js]: Hmm, the misleading message was copied straight from the previous migration - I was under the impression that `warn` output in migrations implied that the operation didn't happen

File: core/server/data/migrations/versions/3.0/2-populate-type-column.js (hunk 5 above)
Comment (allouis) [reply core/server/data/migrations/versions/3.0/2-populate-type-column.js]: I've opted for something which would produce output like 

```
Updating posts, setting "type" column to "page" where "page" column is "true"
```

File: core/server/api/v2/utils/serializers/input/utils/mongo.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
+const _ = require('lodash');
```
Comment (gargol) [inline core/server/api/v2/utils/serializers/input/utils/mongo.js]: Think this module doesn't have to live within the core and can be moved into mongo-utils. It's rather generic and is close by concept to what mongo-utils `expandFilters` function does. Wdyt @allouis? Would also be easier to cover this small piece with a test there :wink:

File: core/server/api/v2/utils/serializers/input/pages.js [hunk 7]
```diff
@@ -3,6 +3,27 @@ const debug = require('ghost-ignition').debug('api:v2:utils:serializers:input:pa
 const converters = require('../../../../../lib/mobiledoc/converters');
 const url = require('./utils/url');
 const localUtils = require('../../index');
+const mongo = require('./utils/mongo');
+
+/*
+ * Replaces references of "page" in filters
+ * with the correct column "type"
+ */
+function replacePageWithType(mongoJSON) {
```
Comment (gargol) [inline core/server/api/v2/utils/serializers/input/pages.js]: Was thinking to write a comment to replace the duplicate structure in here and post serializer, but think the rule of 3 is good to follow :sweat_smile:

Comment (allouis) [reply core/server/api/v2/utils/serializers/input/pages.js]: :sweat_smile: Yeah I think to leave it for now - and see what other changes come about with future api versions!

File: core/server/api/v2/utils/serializers/input/utils/mongo.js (hunk 6 above)
Comment (allouis) [reply core/server/api/v2/utils/serializers/input/utils/mongo.js]: I was thinking the same thing at first, but it felt too specific for mongo-utils :man_shrugging: I'm not too fussed either way - can move it if you like though!

File: core/server/models/plugins/filter.js [hunk 8]
```diff
@@ -61,6 +61,7 @@ const filter = function filter(Bookshelf) {
             let extra = this.extraFilters(options);
             let overrides = this.enforcedFilters(options);
             let defaults = this.defaultFilters(options);
+            let transformer = options.mongoTransformer;
```
Comment (gargol) [inline core/server/models/plugins/filter.js]: Was thinking about the naming of `mongoTransformer` a little and not sure if we want to be specific about "mongo" bit here. For reference to other fields, we have "filter" named as a "filter", not `nqlFilter`, or `defaultFilters` not `defaultNQLFilters` etc. 

In a similar manner would lean towards naming this variable as `filterTransformer`. Wdyt @allouis 	?

Comment (allouis) [reply core/server/models/plugins/filter.js]: Hmmm :thinking: I think `filterTransformer` would imply that the transformer acts on filters defined in the NQL Language (ATM Machine, PIN Number, I know :sweat_smile:) but this works on the Mongo query definition.

I _did_ think that mongo was too much "internals" but the [NQL](https://github.com/NexesJS/NQL) README explicitly states that it is for parsing NQL filters to mongo queries - so I'm not so against it

File: core/server/api/v0.1/posts.js [hunk 9]
```diff
@@ -16,6 +16,55 @@ const Promise = require('bluebird'),
     ],
     unsafeAttrs = ['author_id', 'status', 'authors'];
 
+const mongo = require('../v2/utils/serializers/input/utils/mongo.js');
```
Comment (rishabhgrg) [inline core/server/api/v0.1/posts.js]: Just for naming to be consistent, good to name it as `mongoUtils` as well like the other usage 😄

**Comment by gargol** [comment]
@allouis my understanding is that with v0.1 modifications this PR is meant to land in master instead of v3 branch? If so we'd need a couple more modifications:
1. Rename migrations to suit the version (2.27x ?)
2. There are couple more places in the code I've found that still use the `page` attribute: 
    - https://github.com/TryGhost/Ghost/blob/496f873ac49cd61fd15206252e9fcdf8a7c4f7c6/core/server/services/slack.js#L48 
    - https://github.com/TryGhost/Ghost/blob/496f873ac49cd61fd15206252e9fcdf8a7c4f7c6/core/server/services/xmlrpc.js#L30
3. The whole "frontend" folder needs a review on the use of `page` attribute, especially places like [resource configurations](https://github.com/TryGhost/Ghost/blob/df7e64f/core/frontend/services/routing/config/v0.1.js#L49) that fetch data directly from the db. 

If it's meant to land on the branch though, it's good to go :rocket:

**Comment by allouis** [comment]
@gargol 

1) Rename of the folder deffo needs to happen :+1: - will happen when I know when we're merging.

2) Good catch - will update these :+1: 

3) From what I can see, the only place that hits the DB directly is the url service - which has had the `filter` attribute updated already. I've reviewed the `frontend` folder this morning and but would appreciate a double check before merging :relaxed:

**Comment by gargol** [comment]
@allouis double checked the "frontend" and looks like it's all set. Was thinking some of the model update events that frontend is listening on might be effected (places like [this one](https://github.com/TryGhost/Ghost/blob/df7e64f/core/frontend/services/url/UrlGenerator.js#L198)) but the type/page change doesn't seem to affect any of the code paths. The rest looks good to go I think :+1:

**Pull request merged**